File: tool/validate_courses.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Validate every language's course JSON under assets/courses/.
#
# Checks the manifest wiring, per-course structure, question schema, and
# dictionary coverage of every romanized word a learner can tap. Exits non-zero
# if anything fails, so it can gate a commit.
#
# Usage:
#   ruby tool/validate_courses.rb                    # all languages
#   ruby tool/validate_courses.rb tamil              # one language
#   ruby tool/validate_courses.rb path/to/course.json  # one file, while authoring

require_relative 'courses'

module ValidateCourses
  include Courses

  LEVELS = (5..6).freeze
  QUESTIONS = (8..10).freeze
  OPTION_COUNT = 3

  # Schema 2 is the word course: bare vocabulary against pictures, no
  # sentences. It comes first on the path, because learners were being asked to
  # build sentences out of words nobody had taught them.
  WORD_SCHEMA = 2
  WORD_LEVELS = 5
  WORDS_PER_LEVEL = 12
  WORD_KEYS = %w[concept word gloss].freeze
  REQUIRED_WORD_KEYS = %w[concept word].freeze

  # Courses still being rolled out language by language. Missing is a warning
  # for these; anything else missing is a broken build. Empty this list once
  # every language has one.
  PILOT_COURSES = %w[words].freeze
  NAME_TOKEN = '{name}'
  NAME_ALLOWED_IN = %w[basics introductions].freeze

  MC_PROMPT = 'Choose an appropriate response'
  TRANSLATE_PROMPT = 'Translate the sentence'

  MC_KEYS = %w[type prompt sentence sentenceIsTargetLanguage
               options correctAnswer translatedSentence].freeze
  TRANSLATE_KEYS = (MC_KEYS - ['translatedSentence']).freeze

  module_function

  def check_question(question, where, report, course_id, vocabulary)
    kind = question['type']
    unless %w[multiple_choice translate].include?(kind)
      report.error(where, "unknown type #{kind.inspect}")
      return
    end

    expected_keys = kind == 'multiple_choice' ? MC_KEYS : TRANSLATE_KEYS
    missing = expected_keys - question.keys
    extra = question.keys - expected_keys
    report.error(where, "missing keys #{missing.sort}") unless missing.empty?
    report.error(where, "unexpected keys #{extra.sort}") unless extra.empty?
    return unless missing.empty?

    expected_prompt = kind == 'multiple_choice' ? MC_PROMPT : TRANSLATE_PROMPT
    if question['prompt'] != expected_prompt
      report.error(where, "prompt should be #{expected_prompt.inspect}, got #{question['prompt'].inspect}")
    end

    options = question['options']
    unless options.is_a?(Array) && options.size == OPTION_COUNT
      report.error(where, "expected #{OPTION_COUNT} options, got #{options.is_a?(Array) ? options.size : 0}")
      return
    end
    report.error(where, 'options contain a duplicate') if options.uniq.size != options.size
    unless options.include?(question['correctAnswer'])
      report.error(where, "correctAnswer #{question['correctAnswer'].inspect} is not one of the options")
    end

    report.error(where, 'sentenceIsTargetLanguage must be true') unless question['sentenceIsTargetLanguage'] == true

    sentence = question['sentence']
    report.error(where, 'empty sentence') if sentence.strip.empty?
    report.warn(where, "single-word sentence #{sentence.inspect}") if sentence.split.size < 2

    # The tappable dictionary only covers the target-language strings: the
    # prompt sentence always, plus the options when they are the reply.
    vocabulary.merge(Courses.words_in(sentence))
    if kind == 'multiple_choice'
      options.each { |option| vocabulary.merge(Courses.words_in(option)) }
      report.error(where, 'empty translatedSentence') if question['translatedSentence'].strip.empty?
      if question['correctAnswer'].split.size < 2
        report.warn(where, 'single-word correctAnswer reads like a flashcard')
      end
    elsif options.any? { |option| !option.strip.end_with?('?', '.', '!') }
      # translate options are English meanings.
      report.warn(where, 'an English option is missing end punctuation')
    end

    question.each do |field, value|
      text = value.is_a?(String) ? value : ''
      if text.include?(NAME_TOKEN) && !NAME_ALLOWED_IN.include?(course_id)
        report.error(where, "#{NAME_TOKEN} used outside #{NAME_ALLOWED_IN}")
      end
      report.error(where, "stray '$' in #{field}") if text.include?('$')
    end
  end

  def check_course(path, course_id, report, vocabulary)
    data = Courses.read_json(path)
    where = "#{path.parent.basename}/#{path.basename}"

    if data['course'] != course_id
      report.error(where, "course field is #{data['course'].inspect}, expected #{course_id.inspect}")
    end
    report.error(where, 'missing description') if data['description'].to_s.strip.empty?

    return check_word_course(data, where, report) if data['schema'] == WORD_SCHEMA

    levels = data['levels'] || []
    report.error(where, "#{levels.size} levels, expected #{LEVELS.min}-#{LEVELS.max}") unless LEVELS.cover?(levels.size)

    sentences = Hash.new(0)
    levels.each_with_index do |level, index|
      level_where = "#{where} L#{level['level'] || '?'}"
      if level['level'] != index + 1
        report.error(level_where, "levels must be numbered 1..n, found #{level['level'].inspect} at position #{index + 1}")
      end
      report.error(level_where, 'missing title') if level['title'].to_s.strip.empty?

      questions = level['questions'] || []
      unless QUESTIONS.cover?(questions.size)
        report.error(level_where, "#{questions.size} questions, expected #{QUESTIONS.min}-#{QUESTIONS.max}")
      end

      kinds = questions.group_by { |q| q['type'] }.transform_values(&:size)
      if !questions.empty? && (kinds['multiple_choice'].nil? || kinds['translate'].nil?)
        report.warn(level_where, "level uses only one question type (#{kinds})")
      end

      questions.each_with_index do |question, position|
        check_question(question, "#{level_where} Q#{position + 1}", report, course_id, vocabulary)
        sentences[question['sentence'].to_s] += 1
      end
    end

    sentences.each do |sentence, count|
      report.error(where, "sentence repeated #{count}x: #{sentence.inspect}") if count > 1
    end

    [levels.size, levels.sum { |level| (level['questions'] || []).size }]
  end

  # The shared, language-independent concepts every word course teaches. The
  # English label and the picture live here so an apple is an apple in all
  # thirteen languages; only the target-language word changes.
  def concepts
    @concepts ||= Courses.read_json(Courses::COURSES_DIR / 'concepts.json')['concepts']
                         .to_h { |concept| [concept['id'], concept] }
  end

  # Schema 2: five levels of twelve words, together covering every concept
  # exactly once, each with a picture on disk.
  def check_word_course(data, where, report)
    levels = data['levels'] || []
    unless levels.size == WORD_LEVELS
      report.error(where, "#{levels.size} levels, expected #{WORD_LEVELS}")
    end

    seen = Hash.new(0)
    levels.each_with_index do |level, index|
      level_where = "#{where} L#{level['level'] || '?'}"
      if level['level'] != index + 1
        report.error(level_where,
                     "levels must be numbered 1..n, found #{level['level'].inspect} " \
                     "at position #{index + 1}")
      end
      report.error(level_where, 'missing title') if level['title'].to_s.strip.empty?

      if level.key?('questions')
        report.error(level_where, 'a word level cannot also carry questions')
      end

      words = level['words'] || []
      unless words.size == WORDS_PER_LEVEL
        report.error(level_where, "#{words.size} words, expected #{WORDS_PER_LEVEL}")
      end

      words.each_with_index do |word, position|
        check_word(word, "#{level_where} W#{position + 1}", report)
        seen[word['concept']] += 1
      end
    end

    seen.each do |concept, count|
      report.error(where, "concept #{concept.inspect} taught #{count}x") if count > 1
    end
    missing = concepts.keys - seen.keys
    unless missing.empty?
      report.error(where,
                   "#{missing.size} concepts never taught: #{missing.first(6).join(', ')}")
    end

    # A word course has no questions; the caller only totals them.
    [levels.size, 0]
  end

  def check_word(word, where, report)
    extra = word.keys - WORD_KEYS
    missing = REQUIRED_WORD_KEYS - word.keys
    report.error(where, "unknown keys: #{extra.join(', ')}") unless extra.empty?
    report.error(where, "missing keys: #{missing.join(', ')}") unless missing.empty?
    return unless missing.empty?

    concept = word['concept']
    unless concepts.key?(concept)
      report.error(where, "concept #{concept.inspect} is not in concepts.json")
      return
    end

    art = Courses::IMAGES_DIR / 'vocab' / "#{concept}.svg"
    unless art.exist?
      report.error(where, "no picture for #{concept.inspect}; run tool/fetch_vocab_art.rb")
    end

    target = word['word'].to_s
    report.error(where, 'empty word') if target.strip.empty?
    if target != target.strip
      report.error(where, "word #{target.inspect} has surrounding whitespace")
    end
    # A word course teaches one word, so a space means two of them.
    report.error(where, "word #{target.inspect} contains a space") if target.include?(' ')
    if word.key?('gloss') && word['gloss'].to_s.strip.empty?
      report.error(where, 'gloss is present but empty; omit it to use the concept label')
    end
  end

  def check_non_ascii(path, report, where = nil)
    path.read.each_line.with_index(1) do |line, line_no|
      report.error("#{where || path.basename}:#{line_no}", 'non-ASCII character') unless line.ascii_only?
    end
  end

  def check_language(language, report)
    directory = Courses::COURSES_DIR / language
    manifest_path = directory / 'manifest.json'
    unless manifest_path.exist?
      report.error(language, 'manifest.json is missing')
      return nil
    end

    manifest = Courses.read_json(manifest_path)
    listed = manifest['courses'].map { |course| course['id'] }
    in_tree = manifest['tree'].flatten

    (in_tree - listed).uniq.each do |cid|
      report.error("#{language}/manifest.json", "tree references unknown course #{cid.inspect}")
    end
    (listed - in_tree).uniq.each do |cid|
      report.error("#{language}/manifest.json", "course #{cid.inspect} never appears in the tree")
    end
    manifest['tree'].each do |row|
      unless (1..3).cover?(row.size)
        report.error("#{language}/manifest.json", "tree row of #{row.size} renders nothing (allowed: 1-3)")
      end
    end

    manifest['courses'].each do |course|
      icon = Courses::IMAGES_DIR / "#{course['icon']}.png"
      report.error("#{language}/manifest.json", "icon #{icon.basename} does not exist") unless icon.exist?
      unless course['color'].match?(/\A0x\h{8}\z/)
        report.error("#{language}/manifest.json", "bad colour #{course['color'].inspect}")
      end
    end

    vocabulary = Set.new
    levels = 0
    questions = 0
    listed.each do |cid|
      path = directory / "#{cid}.json"
      unless path.exist?
        # The word course is rolling out one language at a time, on purpose:
        # the shape was proven on the pilot before the other twelve were
        # written. A language that has not got there yet is not a broken build.
        if PILOT_COURSES.include?(cid)
          report.warn(language, "#{cid}.json not written yet")
        else
          report.error(language, "#{cid}.json is missing")
        end
        next
      end
      course_levels, course_questions = check_course(path, cid, report, vocabulary)
      levels += course_levels
      questions += course_questions
    end

    # Non-ASCII would break the romanized-only rule; the manifest is exempt
    # because it carries each language's own name.
    exempt = Courses::NON_COURSE_FILES - ['dictionary.json']
    directory.glob('*.json').sort.each do |path|
      next if exempt.include?(path.basename.to_s)

      check_non_ascii(path, report, "#{language}/#{path.basename}")
    end

    dictionary_path = directory / 'dictionary.json'
    covered = 0
    if dictionary_path.exist?
      entries = Courses.read_json(dictionary_path)
      known = entries.keys.map { |word| Courses.normalize_word(word) }.to_set
      entries.each do |word, gloss|
        report.error("#{language}/dictionary.json", "empty gloss for #{word.inspect}") if gloss.strip.empty?
      end
      vocabulary.delete(Courses.normalize_word(NAME_TOKEN))
      missing = (vocabulary - known).sort
      covered = vocabulary.size - missing.size
      unless missing.empty?
        preview = missing.first(12).join(', ')
        report.error("#{language}/dictionary.json",
                     "#{missing.size} of #{vocabulary.size} words have no gloss: #{preview}" \
                     "#{missing.size > 12 ? ' ...' : ''}")
      end
    else
      report.error(language, 'dictionary.json is missing')
    end

    { language:, courses: listed.size, levels:, questions:,
      vocabulary: vocabulary.size, covered: }
  end

  # Validate one course file on its own — used while authoring, before the rest
  # of a language exists.
  def check_single_file(path, report)
    path = Pathname.new(path)
    unless path.exist?
      report.error(path.to_s, 'file does not exist')
      return
    end
    return if Courses::NON_COURSE_FILES.include?(path.basename.to_s)

    vocabulary = Set.new
    levels, questions = check_course(path, path.basename('.json').to_s, report, vocabulary)
    check_non_ascii(path, report)
    puts "#{path}: #{levels} levels, #{questions} questions, #{vocabulary.size} distinct words"
  end

  def main(argv)
    report = Courses::Report.new

    if argv.first&.end_with?('.json')
      argv.each { |arg| check_single_file(arg, report) }
      report.print_all
      puts report.ok? ? 'OK' : 'FAILED'
      return report.ok? ? 0 : 1
    end

    rows = Courses.languages(argv).filter_map { |language| check_language(language, report) }

    width = rows.map { |row| row[:language].size }.max || 8
    puts format("%-#{width}s  courses  levels  questions  vocabulary", 'language')
    rows.each do |row|
      puts format("%-#{width}s  %7d  %6d  %9d  %5d/%d",
                  row[:language], row[:courses], row[:levels],
                  row[:questions], row[:covered], row[:vocabulary])
    end
    puts "\n#{rows.sum { |row| row[:questions] }} questions across #{rows.size} languages"

    report.print_all
    if report.ok?
      puts "\nall checks passed (#{report.warnings.size} warning(s))"
      0
    else
      puts "\n#{report.errors.size} error(s), #{report.warnings.size} warning(s)"
      1
    end
  end
end

exit ValidateCourses.main(ARGV) if $PROGRAM_NAME == __FILE__

File: tool/validate_courses.py
#!/usr/bin/env python3
"""Validate every language's course JSON under assets/courses/.

Checks the manifest wiring, per-course structure, question schema, and
dictionary coverage of every romanized word a learner can tap. Exits non-zero if
anything fails, so it can gate a commit.

Usage:
    python3 tool/validate_courses.py            # all languages
    python3 tool/validate_courses.py tamil      # one language
"""
import json
import pathlib
import re
import sys
from collections import Counter

ROOT = pathlib.Path(__file__).resolve().parent.parent
COURSES_DIR = ROOT / "assets" / "courses"
IMAGES_DIR = ROOT / "assets" / "images"

MIN_LEVELS, MAX_LEVELS = 5, 6
MIN_QUESTIONS, MAX_QUESTIONS = 8, 10
OPTION_COUNT = 3
NAME_TOKEN = "{name}"
NAME_ALLOWED_IN = {"basics", "introductions"}

MC_PROMPT = "Choose an appropriate response"
TRANSLATE_PROMPT = "Translate the sentence"

MC_KEYS = {"type", "prompt", "sentence", "sentenceIsTargetLanguage",
           "options", "correctAnswer", "translatedSentence"}
TRANSLATE_KEYS = MC_KEYS - {"translatedSentence"}


def normalize_word(word):
    """Mirror of normalizeWord() in lib/courses/course_repository.dart."""
    return re.sub(r"[^a-z0-9]+$", "", re.sub(r"^[^a-z0-9]+", "", word.lower()))


def words_in(sentence):
    return [w for w in (normalize_word(p) for p in sentence.split()) if w]


class Report:
    def __init__(self):
        self.errors = []
        self.warnings = []

    def error(self, where, message):
        self.errors.append(f"{where}: {message}")

    def warn(self, where, message):
        self.warnings.append(f"{where}: {message}")


def check_question(q, where, report, course_id, vocabulary):
    kind = q.get("type")
    if kind not in ("multiple_choice", "translate"):
        report.error(where, f"unknown type {kind!r}")
        return

    expected_keys = MC_KEYS if kind == "multiple_choice" else TRANSLATE_KEYS
    missing = expected_keys - q.keys()
    extra = q.keys() - expected_keys
    if missing:
        report.error(where, f"missing keys {sorted(missing)}")
    if extra:
        report.error(where, f"unexpected keys {sorted(extra)}")
    if missing:
        return

    expected_prompt = MC_PROMPT if kind == "multiple_choice" else TRANSLATE_PROMPT
    if q["prompt"] != expected_prompt:
        report.error(where, f"prompt should be {expected_prompt!r}, got {q['prompt']!r}")

    options = q["options"]
    if not isinstance(options, list) or len(options) != OPTION_COUNT:
        report.error(where, f"expected {OPTION_COUNT} options, got {len(options)}")
        return
    if len(set(options)) != len(options):
        report.error(where, "options contain a duplicate")
    if q["correctAnswer"] not in options:
        report.error(where, f"correctAnswer {q['correctAnswer']!r} is not one of the options")

    if q["sentenceIsTargetLanguage"] is not True:
        report.error(where, "sentenceIsTargetLanguage must be true")

    sentence = q["sentence"]
    if not sentence.strip():
        report.error(where, "empty sentence")
    if len(sentence.split()) < 2:
        report.warn(where, f"single-word sentence {sentence!r}")

    # The tappable dictionary only covers the target-language strings: the
    # prompt sentence always, plus the options when they are the reply.
    vocabulary.update(words_in(sentence))
    if kind == "multiple_choice":
        for option in options:
            vocabulary.update(words_in(option))
        if not q["translatedSentence"].strip():
            report.error(where, "empty translatedSentence")
        if len(q["correctAnswer"].split()) < 2:
            report.warn(where, "single-word correctAnswer reads like a flashcard")
    else:
        # translate options are English meanings.
        if any(not o.strip().endswith(("?", ".", "!")) for o in options):
            report.warn(where, "an English option is missing end punctuation")

    for field, value in q.items():
        text = value if isinstance(value, str) else ""
        if NAME_TOKEN in text and course_id not in NAME_ALLOWED_IN:
            report.error(where, f"{NAME_TOKEN} used outside {sorted(NAME_ALLOWED_IN)}")
        if "$" in text:
            report.error(where, f"stray '$' in {field}")


def check_course(path, course_id, report, vocabulary):
    data = json.loads(path.read_text())
    where = f"{path.parent.name}/{path.name}"

    if data.get("course") != course_id:
        report.error(where, f"course field is {data.get('course')!r}, expected {course_id!r}")
    if not data.get("description", "").strip():
        report.error(where, "missing description")

    levels = data.get("levels") or []
    if not MIN_LEVELS <= len(levels) <= MAX_LEVELS:
        report.error(where, f"{len(levels)} levels, expected {MIN_LEVELS}-{MAX_LEVELS}")

    sentences = Counter()
    for index, level in enumerate(levels, start=1):
        level_where = f"{where} L{level.get('level', '?')}"
        if level.get("level") != index:
            report.error(level_where, f"levels must be numbered 1..n, found {level.get('level')!r} at position {index}")
        if not (level.get("title") or "").strip():
            report.error(level_where, "missing title")

        questions = level.get("questions") or []
        if not MIN_QUESTIONS <= len(questions) <= MAX_QUESTIONS:
            report.error(level_where, f"{len(questions)} questions, expected {MIN_QUESTIONS}-{MAX_QUESTIONS}")

        kinds = Counter(q.get("type") for q in questions)
        if questions and (kinds["multiple_choice"] == 0 or kinds["translate"] == 0):
            report.warn(level_where, f"level uses only one question type ({dict(kinds)})")

        for position, question in enumerate(questions, start=1):
            check_question(question, f"{level_where} Q{position}", report,
                           course_id, vocabulary)
            sentences[question.get("sentence", "")] += 1

    for sentence, count in sentences.items():
        if count > 1:
            report.error(where, f"sentence repeated {count}x: {sentence!r}")

    return len(levels), sum(len(l.get("questions") or []) for l in levels)


def check_language(language, report):
    directory = COURSES_DIR / language
    manifest_path = directory / "manifest.json"
    if not manifest_path.exists():
        report.error(language, "manifest.json is missing")
        return None

    manifest = json.loads(manifest_path.read_text())
    listed = [c["id"] for c in manifest["courses"]]
    in_tree = [cid for row in manifest["tree"] for cid in row]

    for cid in set(in_tree) - set(listed):
        report.error(f"{language}/manifest.json", f"tree references unknown course {cid!r}")
    for cid in set(listed) - set(in_tree):
        report.error(f"{language}/manifest.json", f"course {cid!r} never appears in the tree")
    for row in manifest["tree"]:
        if not 1 <= len(row) <= 3:
            report.error(f"{language}/manifest.json", f"tree row of {len(row)} renders nothing (allowed: 1-3)")

    for course in manifest["courses"]:
        icon = IMAGES_DIR / f"{course['icon']}.png"
        if not icon.exists():
            report.error(f"{language}/manifest.json", f"icon {icon.name} does not exist")
        if not re.fullmatch(r"0x[0-9a-fA-F]{8}", course["color"]):
            report.error(f"{language}/manifest.json", f"bad colour {course['color']!r}")

    vocabulary = set()
    levels = questions = 0
    for cid in listed:
        path = directory / f"{cid}.json"
        if not path.exists():
            report.error(language, f"{cid}.json is missing")
            continue
        course_levels, course_questions = check_course(path, cid, report, vocabulary)
        levels += course_levels
        questions += course_questions

    # Non-ASCII would break the romanized-only rule; the manifest is exempt
    # because it carries each language's own name.
    for path in sorted(directory.glob("*.json")):
        if path.name == "manifest.json":
            continue
        for line_no, line in enumerate(path.read_text().splitlines(), start=1):
            if any(ord(ch) > 127 for ch in line):
                report.error(f"{language}/{path.name}:{line_no}", "non-ASCII character")

    dictionary_path = directory / "dictionary.json"
    covered = 0
    if not dictionary_path.exists():
        report.error(language, "dictionary.json is missing")
    else:
        entries = json.loads(dictionary_path.read_text())
        known = {normalize_word(w) for w in entries}
        for word, gloss in entries.items():
            if not gloss.strip():
                report.error(f"{language}/dictionary.json", f"empty gloss for {word!r}")
        vocabulary.discard(normalize_word(NAME_TOKEN))
        missing = sorted(vocabulary - known)
        covered = len(vocabulary) - len(missing)
        if missing:
            preview = ", ".join(missing[:12])
            report.error(
                f"{language}/dictionary.json",
                f"{len(missing)} of {len(vocabulary)} words have no gloss: {preview}"
                + (" ..." if len(missing) > 12 else ""),
            )

    return {
        "language": language,
        "courses": len(listed),
        "levels": levels,
        "questions": questions,
        "vocabulary": len(vocabulary),
        "covered": covered,
    }


def check_single_file(path, report):
    """Validate one course file on its own — used while authoring, before the
    rest of a language exists."""
    path = pathlib.Path(path)
    if not path.exists():
        report.error(str(path), "file does not exist")
        return
    if path.name in ("manifest.json", "dictionary.json"):
        return  # not a course file; checked as part of the whole language
    vocabulary = set()
    levels, questions = check_course(path, path.stem, report, vocabulary)
    for line_no, line in enumerate(path.read_text().splitlines(), start=1):
        if any(ord(ch) > 127 for ch in line):
            report.error(f"{path.name}:{line_no}", "non-ASCII character")
    print(f"{path}: {levels} levels, {questions} questions, "
          f"{len(vocabulary)} distinct words")


def main():
    args = sys.argv[1:]
    if args and args[0].endswith(".json"):
        report = Report()
        for arg in args:
            check_single_file(arg, report)
        for warning in report.warnings:
            print(f"warning  {warning}")
        for error in report.errors:
            print(f"ERROR    {error}")
        print("FAILED" if report.errors else "OK")
        return 1 if report.errors else 0

    languages = args or sorted(
        p.name for p in COURSES_DIR.iterdir() if p.is_dir()
    )
    report = Report()
    rows = []
    for language in languages:
        stats = check_language(language, report)
        if stats:
            rows.append(stats)

    width = max((len(r["language"]) for r in rows), default=8)
    print(f"{'language':<{width}}  courses  levels  questions  vocabulary")
    for r in rows:
        print(f"{r['language']:<{width}}  {r['courses']:>7}  {r['levels']:>6}  "
              f"{r['questions']:>9}  {r['covered']:>5}/{r['vocabulary']}")
    total = sum(r["questions"] for r in rows)
    print(f"\n{total} questions across {len(rows)} languages")

    for warning in report.warnings:
        print(f"warning  {warning}")
    for error in report.errors:
        print(f"ERROR    {error}")

    if report.errors:
        print(f"\n{len(report.errors)} error(s), {len(report.warnings)} warning(s)")
        return 1
    print(f"\nall checks passed ({len(report.warnings)} warning(s))")
    return 0


if __name__ == "__main__":
    sys.exit(main())
